### DISCUSSION TITLE
How should our build instructions be written?

### DISCUSSION THREAD

**Discussion by thejohnfreeman** [ripple/rippled#4433] [Ideas]
We currently have three open PRs touching our build instructions in a big way:

- #4381 
- #4394 
- #4423 

I think they fundamentally disagree with each other on what the documentation should contain--its **scope**--and how it should be arranged--its **structure**. I think we should resolve that disagreement in this discussion before making further specific changes to the documentation itself. The principles we can establish in this discussion should make it clear exactly what should be written and where it should be placed, including what should be written by others elsewhere and merely linked by us here.

## Audience

The build instructions have several different audiences, divided among several orthogonal dimensions:

- New vs experienced builders of rippled
- Builders for Linux, Mac, and Windows
- Builders using GCC, Clang, MSVC, or another compiler
- Builders using the default options vs customizations (reporting mode, sanitizers, unity build, etc.)

We should serve all audiences, but I think we should make it easy for experienced builders, who will be the most common builders, to find what they're looking for. In my mind, that means putting one-time-only introductory text at the end of BUILD.md, and including copy-pasteable commands for the default options and typical compilers on the major platforms. I don't think it's important to offer the same convenience to inexperienced builders, because they won't be inexperienced for long. I think it is reasonable to expect them to invest a little bit of time up-front to understand how the software is built instead of blindly skimming and copying every command in order.

## Assumptions of reader experience

I would like to assume basic experience with the tools we demonstrate building rippled. We already assume experience with a shell, with CMake, and with how to install software on the reader's system. Our prior documentation included no introduction to CMake teaching readers to how to pass options, or choose a different compiler, or choose a generator. We recently introduced a new tool, Conan. I would like to assume basic familiarity with all of these tools, and to point to external documentation, at most, for those unfamiliar.

It was easy to assume familiarity with a shell and CMake because most people working with C++ _are_ familiar with them. It's not as easy to assume familiarity with Conan because fewer C++ developers are familiar with it. I think it is reasonable to include our own short introduction to Conan, but I would like to make it external to BUILD.md.

## Separate files

I do not want to maintain multiple, mostly identical sets of instructions for different platforms, like we tried in the past. I want one document with instructions for everyone, with minor deviations spelled out. It is easier to maintain, easier to link in answers, and easier for builders building in a new environment to notice the similarities and differences between them.

## Scope

I want to include in BUILD.md:

- Copy-pasteable commands building rippled with default options, using recommended tools on supported platforms.
- Explanations of the different build options for rippled, e.g. static linkage, reporting mode, and sanitizers.
- Steps to troubleshoot common issues specific to building rippled, but not general tool issues like "how to reconfigure CMake" or "how to rebuild a dependency in Conan".
- Platform, compiler, and tool constraints, e.g. minimum versions.

I think there is a place in a separate document for instructions that are not part of building the software, e.g. for how to depend on `xrpl_core` from another CMake project. I think it might be reasonable to include instructions for installing rippled in BUILD.md, but I'd prefer they be separate too.

**Comment by justinr1234** [comment]
Agree with your `Scope` section with 1 caveat:

Common workflows, if expected, should be included. So if it's common to need to know `"how to rebuild a dependency in Conan"`, then we should make it explicit and easy to understand (not require reading all the conan docs to then be able to do this standard operation rippled requires). For people who aren't constant contributors, it makes it tiresome to have to re-familiarize yourself with what rippled does in order to make a small contribution or be able to run a PR.

**Reply by thejohnfreeman** [reply]
I am sympathetic to what you are trying to say, but the line here is very subjective. It is common to need to know how to install a recent version of CMake in one's environment, or a recent version of a compiler (especially now that we're using C++20). We don't include these steps, and I don't think you are suggesting we start. What is the line separating those examples from the one you gave?

**Reply by justinr1234** [reply]
Basically if it's relevant to someone who is new to rippled and something they need to do to run the code in the happy path.

First time:

1. Clone repo
2. Install tools (copy paste)
3. Run build (copy paste)
4. Run tests (copy paste)

Second time after not running the code for a few months and things might be all "mucked up":

1. Pull the latest
2. Reset env or fix common issues (such as rebuilding a dependency) -- copy paste as much as possible
3. Run build
4. Run tests

**Reply by thejohnfreeman** [reply]
Someone new to rippled will need to install CMake. Are you suggesting we include instructions for how to install CMake on the three platforms we support?

**Reply by ckniffen** [reply]
As someone that just compiled C++ code for the first time yesterday that would be helpful.  I used a mixture of the conan docs in this repo and xrpl.org compile docs.

**Reply by thejohnfreeman** [reply]
The document will become incredibly long and 80% non-specific to rippled if we need to include instructions for installing GCC, Clang, Xcode, MSVC, CMake, and Python on all three platforms. Trying to replicate other tools' instructions instead of just pointing to them means our documentation falls out-of-date much faster and more frequently. I would rather just point to the documentation to install CMake, which is what the documentation does currently.

**Reply by oeggert** [reply]
I'm in agreement with @thejohnfreeman on this. Unless there's rippled specific commands/options, we can just link out to the appropriate docs.

My thought is we have an up front "quick start" section. It would contain default commands in one block for people to copy and paste.

Beneath that we have the detailed step-by-step which goes over the nuances and includes links out to external documentation (such as rebuilding dependencies.)

**Reply by justinr1234** [reply]
I don’t think it’s about instructions for installing all of these per se, but calling out a 1 liner copy and paste is valuable. As an example, a lot of windows users don’t know about Chocolatey package manager and it makes installation incredibly easy with a 1 liner. That’s the motivation behind providing a setup script for windows tools, when it makes sense.

Oftentimes in other repos there’s 1 liners for things like `pyenv` to be installed by `brew`. These instructions are quick and easy. To me it’s more about providing a quick on-ramp that leaves a developer thinking “wow that was easy” versus “I need to learn all these pre-reqs before I can even get started”.

Making things that are simple to get into but hard to master I’ve found as very successful.

At the end of the day I advocate for terse instructions that avoid dropping the developer off any cliffs where linking out to other docs might require absorbing too much context that isn’t relevant for the simple task of “I want to compile and run this software”.

Linking out leaves developers to figure out things on their own and also encourages people who wrote the instructions to potentially not validate them on a real machine or a fresh install. I feel like it can encourage “works on my machine”.

“install vscode 2019” leaves out a ton of information about what is truly necessary or even what specific options have been validated by rippled devs. “Install cmake” also leaves out information about specific install options (on windows) that are necessary for rippled. This nuance leaves the developer off a cliff and can lead to discouraging new developers from contributing.

There’s been at least 3 incredibly experienced devs that I know of anecdotally who all independently struggled to compile rippled on various platforms in the last 2 months. This includes C++ experts.

To me, it comes down to reducing instances of struggle while keeping the instructions as terse as possible.

**Reply by drlongle** [reply]
I agree with @justinr1234. I prefer to have simple instructions that "just work". Our potential contributors might be a student or a professional who have a few hours on the weekend. If we make things easy for them, we might have a larger community.

My suggestion would be to have a single document that has
- a section for beginners that contains quick recipes for "popular" choices (e.g. Linux/gcc and Windows/Visual Studio)
- a section for experts that contain build options and further explanation
- a section for troubleshooting
- a section that contain additional information (e.g. links to tutorial for CMake and conan)

It's just a suggestion. I don't feel strongly about it.

**Reply by thejohnfreeman** [reply]
First I want to say that I feel like we're getting closer to identifying the fundamental disconnect among the authors.

It sounds like you want the document to teach people how to set up a basic C++ development environment on their platform. I don't think that is in scope for this document. I think this document should be limited to instructions specific to rippled, and should be able to make a few assumptions about the reader, including that they arrive with a C++ development environment, including a compiler, Git, CMake, and Python.

Once we move to Conan 2.0, the commands to edit the profile disappear and builders will need a text editor. Should we teach them how to install and configure and use Vim? You might think that's crazy, but you look at that starting line the way I look at teaching people to install a C++ development environment.

You mention Chocolatey, but Chocolatey is a very opinionated way to install software that is not recommended by those software authors. Visual Studio does not tell users to use Chocolatey to install Visual Studio. Neither does CMake, Git, or Python. On Windows, I personally prefer to use the official Visual Studio installer for MSVC, and Scoop for everything else. The instructions in BUILD.md are ones we promise to work for readers. I don't want to make that promise for Chocolatey and the rest.

Here's what I propose as a compromise.

- We put a warning at the top of BUILD.md that tells readers that the document assumes they have a C++ development environment, and we include a link to a separate document that walks through how to set one up, for each of the Big Three platforms. One document with all three, not separate documents.
- @justinr1234 and @drlongle can contribute the Windows example, and I can contribute Linux and OSX examples.
- We move the section from BUILD.md titled "A crash course in CMake and Conan" to a separate document, and move the example profile generation command to the section Prerequisites.

This will leave BUILD.md much shorter and easier to navigate, addressing the needs of who I suspect are 90% of visitors: repeat builders of rippled who already have a C++ development environment. 90% of the document remains relevant to those readers, with the 10% being exceptions for specific platforms. The codebase gains instructions that lower the barrier to entry by only assuming the reader knows how to use their computer to install and run software.

Further, I'd like to make these clean up changes:

- Remove the platform-specific build instruction directories under `Builds/`.
- Add a document for post-build operations: installing rippled, depending on xrpl_core, etc.

What do you think?

**Reply by justinr1234** [reply]
Thanks for talking the time, I agree with it being in a separate document. Only thing I’d add is maybe quickly linking out to that document along the way for any main file steps that could be confusing or potentially result in friction for new devs to the project.

**Reply by oeggert** [reply]
So we've got 4 docs planned, then?

1. Set up C++ dev environment.
2. Crash course and set up for CMake and Conan
3. Build Steps/Command
4. Install rippled/other post-build ops

**Reply by drlongle** [reply]
@thejohnfreeman I agree with your suggestion. Personally, I don't think we need to write about Git, CMake and Python. Some quick recipes for the big three platforms are sufficient in my opinion.

**Reply by thejohnfreeman** [reply]
@drlongle I don't think it makes sense to assume the reader knows how to install Git, CMake, and Python, but not a compiler, especially on Windows. I think if we're going to write for a reader who is new to C++ development, we should be complete. In the Linux and Mac instructions, I plan to include instructions for all tools.

@oeggert Mostly. I think it makes sense to separate the instructions for installation and dependence.

- BUILD.md: environmental constraints (e.g. compiler minimum versions), how to build rippled, link to environment.md
- docs/build/
  - environment.md: how to set up a C++ dev environment on the Big Three platforms
  - install.md: how to install rippled
  - depend.md: how to depend on xrpl_core
  - conan.md: crash course

**Reply by thejohnfreeman** [reply]
First phase of these changes has been submitted: https://github.com/XRPLF/rippled/pull/4381#issuecomment-1452817749

**Comment by oeggert** [comment]
I'm pretty much aligned with everything you've outlined here. I think it's better to keep installation and build instructions separate. Were you planning on having an installation md in the repo, or keeping the instructions we have on xrpl.org?

**Reply by thejohnfreeman** [reply]
I'd prefer it live in the repo because it is specific to rippled and not the XRPL, but I don't feel strongly.
